javalib/src/main/scala/java/util/LinkedList.scala

```diff
@@ -0,0 +1,453 @@
+      current.prev.next = node
+      current.prev = node
+
+      _size += 1
+    }
+  }
+
+  override def add(index: Int, element: E): Unit = {
+    checkIndexOnBounds(index)
+    if (index == 0) addFirst(element)
+    else if (index == size) addLast(element)
+    else addNode(getNodeAt(index), element)
+  }
+
+  private def removeNode(node: Node[Box[E]]): E = {
+    if (node.uuid == head.uuid) removeFirst()
+    else if (node.uuid == last.uuid) removeLast()
+    else {
+      var current = node
+
+      current.prev.next = current.next
+      current.next.prev = current.prev
+
+      _size -= 1
+
+      current.value.inner
+    }
+  }
+
+  override   def remove(index: Int): E = {
+    checkIndexInBounds(index)
+    if (index == 0) removeFirst()
+    else if (index == (size - 1)) removeLast()
+    else removeNode(getNodeAt(index))
+  }
+
+  override def indexOf(o: Any): Int = {
+    val boxed = Box(o.asInstanceOf[E])
+    var current: Node[Box[E]] = getNodeAt(0)
+    var count = 0
+    while ((current ne null) &&
+        !current.value.equals(boxed)) {
+      current = current.next
+      count += 1
+    }
+
+    Option(current).fold(-1)(_ => count)
+  }
+
+  override def lastIndexOf(o: Any): Int = {
+    val boxed = Box(o.asInstanceOf[E])
+
+    var current: Node[Box[E]] = getNodeAt(size - 1)
+    var count = _size - 1
+    while ((current ne null) &&
+        !current.value.equals(boxed)) {
+      current = current.prev
+      count -= 1
+    }
+
+    Option(current).fold(-1)(_ => count)
+  }
+
+  def peek(): E =
+    peekFirst()
+
+  def element(): E =
+    getFirst()
+
+  def poll(): E =
+    pollFirst()
+
+  def remove(): E =
+    removeFirst()
+
+  def offer(e: E): Boolean =
+    offerLast(e)
+
+  def offerFirst(e: E): Boolean = {
+    addFirst(e)
+    true
+  }
+
+  def offerLast(e: E): Boolean = {
+    addLast(e)
+    true
+  }
+
+  def peekFirst(): E =
+    if (head.value ne null) head.value.inner
+    else null.asInstanceOf[E]
+
+  def peekLast(): E =
+    if (last.value ne null) last.value.inner
+    else peekFirst()
+
+  def pollFirst(): E =
+    removeFirst()
+
+  def pollLast(): E =
+    removeLast()
+
+  def push(e: E): Unit =
+    addFirst(e)
+
+  def pop(): E =
+    removeFirst()
+
+  private def _removeOccurrence(iter: Iterator[E], o: Any): Boolean = {
+    var changed = false
+    val boxed = Box(o.asInstanceOf[E])
+    while (iter.hasNext() && !changed) {
+      val next = iter.next()
+      if (Box(next) === boxed) {
+        iter.remove()
+        changed = true
+      }
+    }
+
+    changed
+  }
+
+  def removeFirstOccurrence(o: Any): Boolean =
+    _removeOccurrence(iterator(), o)
+
+  def removeLastOccurrence(o: Any): Boolean =
+    _removeOccurrence(descendingIterator(), o)
+
+  override def listIterator(index: Int): ListIterator[E] = {
+    checkIndexOnBounds(index)
+    //new RandomAccessListIterator(this, index, 0, _size)
+    new ListIterator[E] {
+
+      private var last = -1
```

Comment:
`last` is never used except to check whether it is `-1` or not in `checkThatHasLast`. Hence, it can simply be a `hasLast: Boolean`.